### ISSUE TITLE
Autoconfig should take priority over hard binds

### ISSUE THREAD

**Issue by jvook** [libretro/RetroArch#3581]
Currently, autoconfig profiles for controllers will only load if buttons are set to "nul" in retroarch.cfg.  It seems to me that it would be more user-friendly and intuitive to override hard binds with autoconfig profiles.  

I can see there may be arguments against this (for example, maybe in arcade cabinets or retropie installations) but I'm unable to see why the current  behavior would be preferable.

**Comment by dankcushions** [comment]
Autoconfigs are the defaults, though, not an override. They are what the 'Default All' input option loads, in fact. see https://github.com/libretro/RetroArch/issues/2018

**Comment by Ryunam** [comment]
Current behavior allows much more flexibility and freedom in all cases where either the autoconfig is broken or, simply, you want to only change a few binds if compared to the autoconfig default.

**Comment by jvook** [comment]
> Autoconfigs are the defaults, though, not an override. They are what the 'Default All' input option loads, in fact. see #2018

I think this is just semantics.  The fact is that autoconfig profile will not override hard-binded buttons.  This seems backwards.

> Current behavior allows much more flexibility and freedom in all cases where either the autoconfig is broken or, simply, you want to only change a few binds if compared to the autoconfig default.

These issues are easily corrected by saving autoconfig profile if you don't like the default autoconfig.

**Comment by fr500** [comment]
I don't agree with this request.
Autoconf is a method to provide plug and play controller support.  It's not
intended to override settings.

Hard maps should take precedence like they are song now.

**Comment by jvook** [comment]
Plug and play controller support is not working if a user sets an autoconfig.  Plug and play controller support only works with autoconfigs that are included in the distribution.  Otherwise, hardbinds take control, and the user must take the counter-intuitive step of resetting all keys and buttons to default in order to activate the autoconfig that they just saved (not to mention custom key configs are reset in the process).

**Comment by fr500** [comment]
And after such a profile is added plug and play works.
I understand your POV but I don't share it.

The only counterintuitive part of the process is having to hit reset. That
can be fixed but I vote the current behavior should stay.

**Comment by jvook** [comment]
I think we disagree on the meaning of plug and play in this case.

I still see no proferred reason why the current behavior should stay.  If it is some niche use-case, that can be easily rectified by a toggle to force hardbinds take priority, defaulted to off.

In any case, automatically resetting _gamepad buttons only_ after saving autoconfig will result in identical behavior as requested in this issue, and allow the current behavior to continue.

**Comment by jvook** [comment]
Another solution would be a "Create Autoconfig" menu item that creates an autoconfig but does not modify retroarch.cfg

**Comment by fr500** [comment]
Actually after a lot of testing I agree.
Hard maps taking precedence is "safe" but your idea would be more useful to both people with hard maps and autoconf.

I had a reason to ask @twinaphex to make it like this but I really don't remember.

**Comment by Kivutar** [comment]
@fr500 really?

The current behaviour looks perfect to me.

The user plug the joypad:

 * If an autoconfig matches, the jobpad works. DONE.
 * If no autoconfig matches, the user can do a manual bind. And then it works.
 * If the user want to go further, he can save his manual config as an autoconfig.

I don't see how it's a niche use case.

Of course the fact that it still keep the manual bindings in retroarch.cfg is a problem. But it can be resolved easily by using the Bind Default All. Isn't it?

While if we make autoconfig override manual configs, the user loses the ability to override the autoconfig. And they will certainly complain about that.

**Comment by jvook** [comment]
It's the same though in the proposed change:

> The user plug the joypad:
> 
> If an autoconfig matches, the jobpad works. DONE.
> If no autoconfig matches, the user can do a manual bind. And then it works.
> If the user want to **change the default autoconfig**, he can save his manual config as an autoconfig.

Except now the user who is customizing/changing defaults doesn't have to take the strange step of Bind Default All, and losing all his keyboard binds in the process.

There are two possible options to correct this issue:
1.  This proposed change
2.  Save autoconfig command automatically resets _gamepad buttons only_ 

I'm okay with either.

**Comment by jvook** [comment]
> There are two possible options to correct this issue:
> 1. This proposed change
> 2. Save autoconfig command automatically resets gamepad buttons only
> 
> I'm okay with either.

How would use of multiple players affect this?

**Comment by 0x90shell** [comment]
I had to dig deep into the bug vault to find this gem, and I really can't understand the logic here. You've basically broken controller functionality when switching controllers, at least as P1, and when saving your retroarch.cfg. I created a systemctl service to null these bindings so autoconfig always works in the event of an accidental overwrite, but this seems like an extreme course. You could at least provide an option to prioritize autoconfig over retroarch.cfg hardcoded inputs. That way people don't have to go full neckbeard to fix this.

**Comment by 0x90shell** [comment]
Script I'm automating to purge if controller configs accidently get hardcoded to retroarch.cfg

https://gist.github.com/0x90shell/f248cda6e78e763000cb78563c0949f3

**Comment by hizzlekizzle** [comment]
@0x90shell 

> You could at least provide an option to prioritize autoconfig over retroarch.cfg hardcoded inputs. That way people don't have to go full neckbeard to fix this.

Isn't that what the 'reset all to defaults' or whatever option does in settings > input > retropad binds > port X controls? That is, wipe out the manually applied binds so the autoconfig isn't superseded?

**Comment by energypantry** [comment]
Hi! I can take this on with a fast turnaround. Do you accept USDC/USDT (any EVM chain) for payment? If yes, I can start immediately.

**Comment by hizzlekizzle** [comment]
@energypantry  This bounty is no longer active. It died with bountysource.

**Comment by 0x90shell** [comment]
@hizzlekizzle Yes, but I need things to "just work" because my kids shouldn't need to troubleshoot why things randomly break when they use a different controller.

**Comment by hizzlekizzle** [comment]
We already have kiosk mode to prevent messing with settings.

**Comment by 0x90shell** [comment]
In kiosk mode, if the controller is set improperly, how would someone resolve autoconf not working?

**Comment by hizzlekizzle** [comment]
You would resolve that prior to putting it into kiosk mode (or take it out of kiosk mode, fix it and then put it back in). Either you trust your users to change settings or you don't.

**Comment by sonninnos** [comment]
While this won't solve the initial issue, since it is not going to happen as-is, it will surely help somewhat: https://github.com/libretro/RetroArch/pull/18738

**Comment by 0x90shell** [comment]
> You would resolve that prior to putting it into kiosk mode (or take it out of kiosk mode, fix it and then put it back in). Either you trust your users to change settings or you don't.

Yea, that is not conducive to a pickup and play experience with kids. I need autoconf, which I've already tinkered with, to work without the need for others to go under the hood with settings.
